payment/app.py

```diff
@@ -1,87 +1,113 @@
         return {'Error': 'User not found'}, 404
-    return {"user_id": int(user_id), "credit": int(user[0])}, 200
+    return {'user_id': user_id, 'credit': user['credit']}, 200
 
 
 @app.post('/add_funds/<user_id>/<amount>')
 def add_credit(user_id: str, amount: int):
     if int(amount) < 0:
         return {'Error': 'Amount must be positive'}, 400
-    if not db.exists(f'user:{user_id}'):
-        return {'Error': 'User not found'}, 400
-    db.hincrby(f'user:{user_id}', 'credit', int(amount))
-    return 'Success', 200
+    user = user_collection.find_one_or_404({'_id': ObjectId(user_id)})
+    credit: int = int(user['credit'])
+    newCredit = credit + int(amount)
+    result = user_collection.update_one(
+        # Filter to check if the credit is unchanged
+        {'_id': ObjectId(user_id), 'credit': { '$eq': credit}},
+        {'$set': {'credit': newCredit}}
+    )
+
+    if result.modified_count != 1:
+        return {'Error': 'The credit was updated by someone else'}, 400
+    
+    return {'Success': 'Credit is updated successfully'}, 200
+
 
 
 @app.post('/pay/<user_id>/<order_id>/<amount>')
 def remove_credit(user_id: str, order_id: str, amount: int):
-    credit = db.hget(f'user:{user_id}', 'credit')
-    if int(credit) < int(amount):
-        return 'Not enough credit', 400
-    order = db.exists(f'paid_orders:{order_id}')
+    
+    order = paid_order_colleciton.find_one({'order_id': order_id})
 
     if order:
         return 'Order already paid', 401
 
-    p = db.pipeline(transaction=True)
-    p.hincrby(f'user:{user_id}', 'credit', -int(amount))
-    p.hmset(f'paid_orders:{order_id}', {
-            'amount_paid': amount, 'user_id': user_id})
-    p.execute()
+    user = user_collection.find_one_or_404({'_id': ObjectId(user_id)})
+    credit: int = int(user['credit'])
+    
+    # For some reason amount is not an int but string at this point
+    if credit < int(amount):
+        return 'Not enough credit', 400
+
+    # Update to transaction
+    newCredit = credit - int(amount)
+    result = user_collection.update_one(
+        # Filter to check if the credit is unchanged
+        {'_id': ObjectId(user_id), 'credit': { '$eq': credit}},
+        {'$set': {'credit': newCredit}}
+    )
+
+    if result.modified_count != 1:
+        return {'Error': 'Something went wrong checking out the order'}, 400
+
+    paid_order_colleciton.insert_one({'order_id': order_id, 'amount': amount}) 
     return 'Success', 200
 
 
 @app.post('/cancel/<user_id>/<order_id>')
 def cancel_payment(user_id: str, order_id: str):
-    order = db.hmget(f'paid_orders:{order_id}', 'amount_paid')
-    if None in order:
-        return {'Error': 'Order not found'}, 404
-
-    amount = order[0]
-    p = db.pipeline(transaction=True)
-    p.hincrby(f'user:{user_id}', 'credit', amount)
-    p.delete(f'paid_orders:{order_id}')
-    p.execute()
+    order = paid_order_colleciton.find_one_or_404({'order_id': order_id})
+    amount = int(order['amount'])
```

Comment:
Will this crash if order is 404?

Reply:
Yes, however I updated the code to add a custom error message instead of a generic page